applications/StructuralMechanicsApplication/custom_utilities/large_displacement_variables.h

```diff
@@ -0,0 +1,157 @@
+
+    void CalculateDetF(std::size_t IntegrationIndex, bool IsAxisymmetric = false);
+
+    void CalculateDN_DX0(std::size_t IntegrationIndex);
+
+    void CalculateAxisymmetricF(std::size_t IntegrationIndex);
+
+    GeometryType const& mrGeom;
+    Matrix mJ;
+    Matrix mInvJ0;
+    double mDetJ0;
+    Matrix mF;
+    Matrix mDN_DX0;
+    double mDetF;
+    std::size_t mInternalIntegrationIndex = -1;
+    std::size_t mIntegrationIndexF = -1;
+    std::size_t mIntegrationIndexDN_DX0 = -1;
+    std::size_t mIntegrationIndexDetF = -1;
+};
+
+/**
+ * @class LargeDisplacementKinematicVariables
+ * @ingroup StructuralMechanicsApplication
+ * @brief Kinematic variables for large displacements with reference configuration.
+ * @details Provides strain-related fields.
+ */
+class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) LargeDisplacementKinematicVariables
+{
+public:    
+    LargeDisplacementKinematicVariables(LargeDisplacementDifferentialVariables& rDiffVars,
+                                        bool IsAxisymmetric = false);
+
+    const Matrix& StrainTensor(std::size_t IntegrationIndex);
+
+    /* Return mutable vector for compatibility with ConstitutiveLaw. */
+    Vector& StrainVector(std::size_t IntegrationIndex);
+
+    const Matrix& B(std::size_t IntegrationIndex);
+
+private:
+    LargeDisplacementDifferentialVariables& mrDiffVars;
+    bool mIsAxisymmetric;
+    Matrix mStrainTensor;
+    Vector mStrainVector;
+    Matrix mB;
+    std::size_t mCurrentIntegrationIndex = -1;
+    std::size_t mStrainVectorIntegrationIndex = -1;
+
+    void Recalculate(std::size_t IntegrationIndex);
+
+    void CalculateStrainTensor(std::size_t IntegrationIndex);
+
+    void CalculateB(std::size_t IntegrationIndex);
+
+    void CalculateB_Axisymmetric(std::size_t IntegrationIndex);
+};
+
+/**
+ * @class LargeDisplacementKinematicVariables
+ * @ingroup StructuralMechanicsApplication
+ * @brief Variables related to large displacement deformations.
+ * @details @see LargeDisplacementDifferentialVariables, @see LargeDisplacementKinematicVariables.
+ */
+class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) LargeDisplacementDeformationVariables
+{
+public:
+    typedef Geometry<Node<3>> GeometryType;
+
+    LargeDisplacementDeformationVariables(GeometryType const& rGeom,
+                                          bool IsAxisymmetric = false);
+
+    const Matrix& F(std::size_t IntegrationIndex);
+
+    double DetF(std::size_t IntegrationIndex);
+
+    const Matrix& DN_DX0(std::size_t IntegrationIndex);
+
+    double DetJ0(std::size_t IntegrationIndex);
+
+    const Matrix& StrainTensor(std::size_t IntegrationIndex);
+
+    Vector& StrainVector(std::size_t IntegrationIndex);
+
+    const Matrix& B(std::size_t IntegrationIndex);
+
+private:
+    LargeDisplacementDifferentialVariables mDiffVars;
+    LargeDisplacementKinematicVariables mKinVars;
+    const bool mIsAxisymmetric;
+};
+
+void CalculateB_2D(Matrix const& rF, Matrix const& rDN_DX0, Matrix& rB);
```

Comment:
This should be done by the element. For a B-bar element for example this is fundamental...

Reply:
I will reply to this and the F both at the end.